examples/Spec/TicTacToe.hs

```diff
@@ -0,0 +1,519 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Spec.TicTacToe (ticTacToeGenTests,ticTacToeGenSelfTests) where
+import Proper.HasLogicalModel
+import Proper.LogicalModel
+import Proper.HasParameterisedGenerator
+import Proper.HasPermutationGenerator
+import SAT.MiniSat ( Formula (..) )
+import Hedgehog (Gen)
+import qualified Hedgehog.Gen as Gen
+import Hedgehog.Range (linear,singleton)
+import Test.Tasty (TestTree,testGroup)
+import Test.Tasty.Hedgehog (fromGroup)
+import qualified Data.Set as Set
+import Data.Maybe (isNothing)
+import Data.Proxy (Proxy(..))
+
+data Player = X | O deriving stock (Eq,Show)
+
+data TicTacToeMove =
+  TicTacToeMove {
+    from :: [Maybe Player]
+  , to   :: [Maybe Player]
+  , player :: Player
+  , declare :: Bool
+  } deriving stock (Show)
+
+data TicTacToeProperty =
+      FromBoardIsCorrectSize
+    | FromBoardIsEmpty
+    | FromBoardInInitialState
+    | FromBoardHasOddNumberOfPieces
+    | FromBoardHasEqualNumberOfPieces
+    | FromBoardHasOneMoreXThanO
+    | ToBoardIsCorrectSize
+    | ToBoardIsEmpty
+    | ToBoardEqualToFromBoard
+    | PlayerIsX
+    | IsPlayersTurn
+    | WinDeclared
+    deriving stock (Eq,Ord,Enum,Show,Bounded)
+
+instance LogicalModel TicTacToeProperty where
+  logic = All [Var FromBoardInInitialState :<->: (All $ Var <$> [FromBoardIsCorrectSize
+                                                                ,FromBoardIsEmpty
+                                                                ])
+              ,Var FromBoardHasOddNumberOfPieces :->: Not (Var FromBoardIsEmpty)
+              ,Var IsPlayersTurn :<->:
+                   ((Var PlayerIsX :&&: (Not $ Var FromBoardHasOddNumberOfPieces))
+                 :||: ((Not $ Var PlayerIsX) :&&: (Var FromBoardHasOddNumberOfPieces)))
+              , Var FromBoardIsEmpty :->: Var FromBoardHasEqualNumberOfPieces
+              , Var FromBoardHasEqualNumberOfPieces :->: (Not $ Var FromBoardHasOddNumberOfPieces)
+              , Var FromBoardHasOneMoreXThanO :->: Var FromBoardHasOddNumberOfPieces
+              , Var ToBoardEqualToFromBoard :->: (All [
+                   Var FromBoardIsCorrectSize :<->: Var ToBoardIsCorrectSize
+                 , Var FromBoardIsEmpty :<->: Var ToBoardIsEmpty
+                                                      ])
+              , (All $ Var <$> [FromBoardIsCorrectSize
+                               ,FromBoardIsEmpty
+                               ,ToBoardIsCorrectSize
+                               ,ToBoardIsEmpty]) :->: Var ToBoardEqualToFromBoard
+              ]
```

Comment:
Here is an unfinished example of a more complex model with a generator written using `HasPermutationGenerator`. The `PermutationGenerator` self-test test suite that checks correctness of the `PermutationGenerator` morphisms runs around 1500 tests in 40 seconds which is fast enough to give compiler like feedback during development.